FormField should autovalidate only if its content was changed

pedromassango opened the pull request:
## Description

I propose to auto-validate `FormField` only if its content was changed since its prevents end users to see validation errors even without changing the field's content.

## Related Issues

Fixes: https://github.com/flutter/flutter/issues/56363
Fixes: https://github.com/flutter/flutter/issues/18885
Fixes: https://github.com/flutter/flutter/issues/15404
Fixes: https://github.com/flutter/flutter/issues/36154
Fixes: https://github.com/flutter/flutter/issues/48876

## Tests

I added the following test cases:


- [x] Do not auto-validate before value changes if widget.autovalidateMode (new) equals to  AutovalidateMode.onUserInteraction;
- [x] auto-validate after value changes if widget.autovalidateMode equals to AutovalidateMode.onUserInteraction;
- [x] Form only validates form fields after one of the form fields changes if widget.autovalidateMode is AutovalidateMode.onUserInteraction;
- [x] auto-validate before value changes if autovalidate is true and widget.autovalidateMode is AutovalidateMode.always;
- [x] Form auto validate form fields even if none of them changes if widget.autovalidateMode is always;
autovalidate parameter is still used if true;
- [x] DropdownButtonFormField passes autoValidateMode to  superclass;
- [x] Form.reset() reset form fields and auto validation will only happen on the next user interaction if autoValidateMode is onUserInteraction;

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I signed the [CLA].
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. If not, delete the remainder of this section.
    - [x] I wrote a design doc: https://docs.google.com/document/d/16HAURcOtYAoHntDOEtdGy9KkKSItulG9Ldyjzdaosqo/edit?usp=sharing
    - [x] I got input from the developer relations team, specifically from: @chunhtai 
    - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes

pedromassango commented:
@TahaTesser could you please hep me figure it out why the tests are not passing? I didn't touch any file related to these tests

SpajicM commented:
Duplicate of https://github.com/flutter/flutter/pull/56132?

Ahmadre commented:
> Duplicate of #56132?

wait a minute 🤔 could it be, that this PR is more up-to-date then the one you're referencing to?
I mean this:

https://github.com/flutter/flutter/pull/56365/commits/6c89e805f84a68410962738ed7d5081c8118f2f3#diff-fe715f72ed581d9662741a24d2d98e45R437-R438

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,8 +434,8 @@ class FormFieldState<T> extends State<FormField<T>> {
-    // Only autovalidate if the widget is also enabled
-    if (widget.autovalidate && widget.enabled)
+    // Only autovalidate if the widget is also enabled and if its content has changed
+    if (widget.autovalidate && widget.enabled && _valueChanged)
```
what if validator has changed?

chunhtai replied:
instead of using the _valueChanged you can cache the last validated value and only validate when the value is different.

pedromassango replied:
But if I cache the last value (lets say the first value that is a empty string) then we will not be able to validate if the value is a empty string.

pedromassango replied:
Working on this

chunhtai replied:
I see, so the use case is when form first build we don't want to validate until user touches it. but then once the user touches it, we want to start validate even if the form is still empty after user touched it.

chunhtai replied:
I think you also need to somehow set _valueChanged back to false after validation in this current implementation.

pedromassango replied:
Why we need to say "value was not changed" after each validation?

chunhtai replied:
Sorry I was confused a bit at the name valueChanged, I thought it meant the value has changed since last validation. it should be something like stateIsDirty or startsValidation. There is another PR tries to achieve the same thing https://github.com/flutter/flutter/pull/56132.

In this case the validator change does not really matter here.

pedromassango replied:
Right

chunhtai reviewed (commented):
Tests are failing in packages/flutter/test/material/text_form_field_test.dart

chunhtai commented:
This also needs test

goderbauer commented:
/cc @justinmc

pedromassango commented:
I'm trying to fix the tests but I come with one use-case: if the user provide a default value (lets say non-empty string) should we validate that:

- on the first time
- once the TextField gains focus and

I think the second approach is better to fix this issue, but then we get another issue: what if the user don't even touch the TextField?!

This is causing this test to fail: https://github.com/flutter/flutter/blob/f865ac7e259d320d7e565f887f894f293a463782/packages/flutter/test/material/dropdown_form_field_test.dart#L151

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -414,6 +417,7 @@ class FormFieldState<T> extends State<FormField<T>> {
   @protected
   void setValue(T value) {
     _value = value;
+    _valueChanged = true;
```
This method is called by subclass textformfield.didChangeWidget.

It means the auto validate will be enabled if the textformfield was built twice with different text editing controller.

pedromassango replied:
Right, so I think I can remove here right?

justinmc reviewed (commented):
~~What if a developer wants to validate when the overall form is submitted, and the user submits the form but never focused or changed a field?  I think it unexpectedly wouldn't show an error message for the field.  That case should be handled.~~ I checked out this branch and tried this, and it seems to work.  `validate` is called directly, so this doesn't prevent the errors from being shown.

I mentioned this in the other PR https://github.com/flutter/flutter/pull/56132, but strictly speaking this is a breaking change.  I agree that very few people probably want to validate immediately before the field has even been focused, though.  After this change, if anyone does want that behavior, it would be possible to do by calling validate manually on first render.  @HansMuller do you have any opinions on making this breaking change?  Assuming the case in my first paragraph above is fixed.

If we are going to go forward with this change, @pedromassango and @ashank96 need to decide who's PR we're going with.  This PR probably should have its own test like https://github.com/flutter/flutter/pull/56132 does.

pedromassango commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,8 +434,9 @@ class FormFieldState<T> extends State<FormField<T>> {
-    // Only autovalidate if the widget is also enabled
-    if (widget.autovalidate && widget.enabled)
+    // Only autovalidate if the widget is also enabled and
+    // if its content has changed or has focus
```
it will validate the TextField! I'm just preventing autovalidation. so calling

` formKey.currentState.validate();` is working with no problems :)

jmagman committed:
Make initial daemon devices population fast (#56959)

engine-flutter-autoroll committed:
Roll engine b59e3e9c39a2..ae2222f47e78 (8 commits) (#57150)

* 7b6406720 Use 'message' as the parameter name in FlMessageCodec::encode_message (flutter/engine#18253)

* 429beae8a Roll src/third_party/skia 3d2c41b773f6..3ebadcc98eab (14 commits) (flutter/engine#18333)

* 66ba3a748 Roll src/third_party/dart 2bf325900586..d6fed1f62444 (1 commits) (flutter/engine#18334)

* 2f8495a5e Completely disable paving the device on Fuchsia (flutter/engine#18340)

* 006dbfce2 Always keep thread merged when there are platform views. (flutter/engine#18245)

* 21b4d2f1f [web] Fix paragraph positioning (flutter/engine#18329)

* efdc0994a Re-enable Fuchsia tests (flutter/engine#18342)

* ae2222f47 Revert "Re-enable Fuchsia tests (#18342)" (flutter/engine#18345)

Piinks committed:
Step 1 of 3: Add opt-in for debugCheckHasMaterialLocalizations assertion on TextField (#56090)

darrenaustin committed:
Localized new strings added in the redesigned Material Date Picker (#56645)

christopherfujino committed:
throw more specific toolexit when git fails during upgrade (#57162)

ferhatb committed:
Unskip TextStyle web tests (#56710)

Piinks committed:
Remove deprecated child parameter (#57065)

darrenaustin committed:
Updated gen_missing_localizations to copy the english strings instead of using 'TBD' for other locales. (#57053)

Commit:
[flutter_tools] do not set timestamp of package_config file (#57077)

LongCatIsLooong committed:
remove redundant transform (#57085)

Commit:
[flutter_tools] expose track-widget-creation to build aar (#57117)

chunhtai committed:
update initial route documentation (#57136)

Commit:
[flutter_tools] ensure package_config is re-created if pub get is run (#57184)

justinmc committed:
Bring back paste button hide behavior 3 (#57139)

bkonyi committed:
Disable DartDev when launching flutter_tools (#57143)

shihchanghsiungsonos committed:
[Add2App Android] Fix the issue of Hotreload broken on latest Dev release with Android device (#57145)

Commit:
[flutter_tools] allow adb to fail to unforward without crashing (#57173)

Commit:
[flutter_tools] fix period in URL for androidX incompat (#57182)

Commit:
[null-safety] disable tests until framework has migrated (#57235)

ferhatb committed:
[web] Add path construction benchmark (#57016)

* Add path construction benchmark

* Fix analyzer warnings

* Split file into test+generated code. Add comment

* Add license to new file

* remove unused import

pedromassango commented:
@chunhtai Could you please inform the current state of this PR to be merged? I really need this. Sorry to bother you.

liyuqian committed:
Rename performance.md to performance_speed.md

So we can have another template for non-speed performance issues (e.g., memory, app size, memory).

liyuqian committed:
Create performance_others.md

liyuqian committed:
Update performance_others.md

zanderso committed:
[flutter_tools] Delete system temp entries on fatal signals (#55513)

dkwingsmt committed:
Animation sheet recorder (#55527)

Jjagg committed:
Fix typo in 'flutter create --list-samples' help (#57005)

darrenaustin committed:
Honor the InputDecoratorTheme in the text input fields used by the Date Pickers. (#57189)

creativecreatorormaybenot committed:
Fix docs (#57195)

engine-flutter-autoroll committed:
Roll engine ae2222f47e78..47513a70eb6d (21 commits) (#57252)

* 9e3e3ba68 [web] Represent CSS identity transforms as 'none' instead of null (flutter/engine#18288)

* 80fa77eb3 Roll src/third_party/skia 3ebadcc98eab..056d543c91e0 (8 commits) (flutter/engine#18344)

* 480d8e458 Fix scenario platform view tests on Android (flutter/engine#18332)

* 03856642d Roll src/third_party/dart d6fed1f62444..29c00e28f350 (16 commits) (flutter/engine#18347)

* 8371b44e5 Roll src/third_party/skia 056d543c91e0..71903997254f (7 commits) (flutter/engine#18350)

* f321613e5 Add guards on FlValue methods to check for NULL values (flutter/engine#18226)

* dc93db5ab Move FlutterLoader disk I/O to a background thread to comply with Android strict mode (flutter/engine#18241)

* bf1287c12 null-annotate lerp.dart, annotations.dart, channel_buffers.dart, hash_codes.dart (flutter/engine#18348)

* 960035462 Add FlBasicMessageChannel (flutter/engine#18189)

* df2dface3 Roll src/third_party/skia 71903997254f..6c3db04c8b03 (9 commits) (flutter/engine#18361)

* 5ad4f9ee4 Add FlJsonMessageCodec (flutter/engine#18221)

* 84ea89262 Roll src/third_party/skia 6c3db04c8b03..7156db260239 (4 commits) (flutter/engine#18370)

* f848069d4 Roll src/third_party/dart 29c00e28f350..f99631b12c4a (29 commits) (flutter/engine#18373)

* c791ce9b1 Roll src/fuchsia/sdk/mac from gOhJW... to Vepm4... (flutter/engine#18377)

* 5b62a6351 Roll src/third_party/dart f99631b12c4a..e0257265d34e (2 commits) (flutter/engine#18378)

* 0b4100946 Roll src/third_party/skia 7156db260239..5b2ede3d0d44 (8 commits) (flutter/engine#18380)

* 08b61ce56 Delete unused decode UTF-8, JSON functions (flutter/engine#18360)

* 73d835cbd Roll src/third_party/dart e0257265d34e..2676764792b2 (4 commits) (flutter/engine#18383)

* 9e166fbf4 Roll src/third_party/skia 5b2ede3d0d44..39ec60aa8348 (5 commits) (flutter/engine#18384)

* 2cdbc7f92 Remove pipeline in favor of layer tree holder (flutter/engine#18285)

* 47513a70e Roll src/third_party/skia 39ec60aa8348..79c5674a4ca1 (3 commits) (flutter/engine#18389)

yjbanov committed:
Allow Recorder override shouldContinue (#57039)

dkwingsmt committed:
MouseCursor uses a special class instead of null to defer (#57094)

* Uses a special cursor value MouseCursor.defer to mark deferring. MouseTrackerAnnotation.cursor is now non-null. The mouseCursor argument of widgets or render objects can be null, which indicates using the default value.
* Moves SystemMouseCursors.uncontrolled to MouseCursor.uncontrolled.
* Changes how MouseCursor.debugDescription is defined. Previously MouseCursor.toString returns $runtimeType($debugDescription), while now it returns $debugDescription. Implementations of classes are updated, except for the ones of MouseCursor.defer and MouseCursor.uncontrolled are simply "defer" and "uncontrolled".

JoseAlba committed:
Added Dartpad and Image examples to Slider and RangeSlider docs (#57047)

engine-flutter-autoroll committed:
50e55cf69 Roll src/fuchsia/sdk/mac from Vepm4... to 61d8Z... (flutter/engine#18391) (#57258)

chaselatta committed:
add missing deps to flutter_test BUILD.gn (#57270)

Commit:
Revert " Bring back paste button hide behavior 3 (#57139)" (#57286)

This reverts commit 8de07d5527bcdc6b02e43e8efed19219a84bf82e.

chunhtai commented:
I leave a similar comment on other pr,
can this also apply to the form.autovalidate which is doing the same thing except it only validate after the second build.

pedromassango commented:
> I leave a similar comment on other pr,
> can this also apply to the form.autovalidate which is doing the same thing except it only validate after the second build.

I will try to reproduce the issue with the `Form` widget and see what I can do about it

pedromassango commented:
> I leave a similar comment on other pr,
> can this also apply to the form.autovalidate which is doing the same thing except it only validate after the second build.

@chunhtai I wa able to reproduce this. Do you think that I must open a new issue and PR or just fix it on this one?

chunhtai commented:
I will suggest implement it in the same pr so that it will be easier to revert in case somethings go wrong and does not end up with form.autovalidate behavior differently than formfield.autovalidate

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import '../../widgets.dart';
```
This import is weird, you should directly import only the file needed in the same directory

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -337,6 +338,7 @@ class FormField<T> extends StatefulWidget {
 class FormFieldState<T> extends State<FormField<T>> {
   T _value;
   String _errorText;
+  bool _valueChanged = false;
```
per our previous discussion, would you rename this to be more descriptive? such as _hasInteractedByUser

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,8 +434,9 @@ class FormFieldState<T> extends State<FormField<T>> {
-    if (widget.autovalidate && widget.enabled)
+    // Only autovalidate if the widget is also enabled and
+    // if its content has changed or has focus
+    if (widget.autovalidate && widget.enabled && (_valueChanged || Focus.isAt(context)))
```
Can you remind me why do we need Focus.isAt(context)?

pedromassango replied:
I can't tell you right now. Le me investigate.

pedromassango replied:
@chunhtai I added this to make this test pass : https://github.com/flutter/flutter/blob/749a30b4380761d327a70df99fc18f2d9f68c99c/packages/flutter/test/material/dropdown_form_field_test.dart#L151

It was to assert that the validation of `DropdownButtonFormField` only happens when it is focused since there the user does not type the text right in the Field.

It seems unnecessary to me right now, since I can just use the value state changes property

pedromassango committed:
Merge branch 'patch-1' of https://github.com/pedromassango/flutter

pedromassango committed:
fix import directive

pedromassango committed:
Improved variable name

pedromassango committed:
Fix test case

pedromassango committed:
remove unnecessary focus verification

pedromassango committed:
removed unused import

pedromassango committed:
added test to assert that Form is auto validating without user iteration

pedromassango committed:
Making test pass

pedromassango commented:
@chunhtai please take a look at the latest changes

pedromassango committed:
Remove trailing spaces

coin3x committed:
Make CircularProgressIndicator's animation match native (#50412)

engine-flutter-autoroll committed:
Roll engine 50e55cf69e8a..aaf9e79f1d29 (2 commits) (#57269)

jonahwilliams committed:
[flutter_tools] surface missing assets originating package (#55701)

enolcasielles committed:
avoid AnimatedBuilder rebuilding widget (#55832)

jonahwilliams committed:
[flutter_tools] cache-bust in service worker (#56786)

liyuqian committed:
Remove obsolete UpdateCounted prefix (#57167)

Hixie committed:
Add more documentation to addTimingsCallback (#56952)

zanderso committed:
Update packages (#57321)

Commit:
Add option for ExpansionTile to maintain the state of its children when collapsed (#57172)

stuartmorgan-g committed:
Switch to CMake for Linux desktop (#57238)

Updates the Linux templates to use CMake+ninja, rather than Make, and updates the tooling to generate CMake support files rather than Make support files, and to drive the build using cmake and ninja.

Also updates doctor to check for cmake and ninja in place of make.

Note: While we could use CMake+Make rather than CMake+ninja, in testing ninja handled the tool_backend.sh call much better, calling it only once rather than once per dependent target. While it does add another dependency that people are less likely to already have, it's widely available in package managers, as well as being available as a direct download. Longer term, we could potentially switch from ninja to Make if it's an issue.

Fixes #52751

renefloor committed:
correctly dispose listeners by image widget (#57201)

a14n committed:
add @factory to create* methods (#57299)

feinstein committed:
Fixed a typo. (#57412)

Zazo032 committed:
Fix Autofill example (#57319)

pedromassango committed:
Add autofill support for TextFormField (#57332)

flar committed:
Protect the deletion of the local engine temp dir in case it is already deleted (#57345)

engine-flutter-autoroll committed:
Roll engine aaf9e79f1d29..2bd71fb60a43 (52 commits) (#57402)

HansMuller committed:
ThemeData.brightness == ThemeData.colorScheme.brightness (#56956)

engine-flutter-autoroll committed:
Roll engine 2bd71fb60a43..aafd9f72283f (9 commits) (#57473)

Hattomo committed:
fix typo (#57487)

AyushBherwani1998 committed:
Add ExpansionTile.childrenPadding property (#57291)

pingbird committed:
Make _RenderButtonBarRow.constraints null aware (#57261)

jonahwilliams committed:
[flutter_tools] check for Runner.sln when parsing for plugins (#57392)

Throw a toolExit if the windows plugin logic runs on an invalid windows project. Update the supported project check to validate the existence of a Runner.sln file

stuartmorgan-g committed:
Fix CMake invocation for 3.10 compat (#57415)

The -S/-B format of specifying source and build directories wasn't added
until 3.13. For 3.10 compatibility, use the older form of having the
working directory as the build directory, and the last argument as the
source directory.

Fixes #57409

stuartmorgan-g committed:
Add Linux GTK artifacts to unpack list (#57452)

To simplify development of the in-progress GTK embedding, which will
replace the GLFW embedding on Linux, add the GTK artifacts to the unpack
list. This means that until the transition, both the GLFW and GTK
artifacts will be unpacked.

Part of #54860

pennzht committed:
Update flutter_gallery_assets to ^0.2.0 (#57328)

* Bump flutter_gallery_assets version to ^0.2.0.

* Run `flutter update-packages --force-upgrade`.

* Add comment.

* Remove duplicate key.

stuartmorgan-g committed:
Temporarily allow pluginClass: none on desktop (#57498)

Treats 'pluginClass: none' as equivalent to having no 'pluginClass'
entry on the desktop platforms, to satisy stable channel plugin
validation of Dart-only desktop plugin implementations. See
issue for full details.

Part of https://github.com/flutter/flutter/issues/57497

jonahwilliams committed:
[flutter_tools] minor cleanups to try catch (#57446)

- OsError now implements exception, remove workaround.
- Prefer is check to catching TypeError

jonahwilliams committed:
[flutter_tools] put system clock on globals (#57447)

ferhatb committed:
Resolve skip (#57240)

engine-flutter-autoroll committed:
Roll engine aafd9f72283f..90f45bd3efce (4 commits) (#57499)

JoseAlba committed:
Value Indicator uses Global position (#57327)

* Value Indicator uses Global position when calculating horizontal position

* fixed test

* fixed analyzer

* Made globalCenter variable and added comments

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -163,7 +163,8 @@ class FormState extends State<Form> {
   Widget build(BuildContext context) {
-    if (widget.autovalidate)
+    // validate only if one of the text form fields changes
+    if (widget.autovalidate && _generation > 0)
```
I think this is more complex than i thought, the reset method will not reset the _generation but increment it by 1. The documentation of that method actually mentioned the field will gets revalidate if autovalidate is true, but that is contradict with the formfield.reset.

pedromassango replied:
This make sense. 

I think form.reset should not validate de formFields, this may cause what we are trying to fix: validate the without user iteraction

chunhtai reviewed (commented):
Since it is getting more complex, have you thought about write a design doc that summarize the new behavior so that people can quickly understand what is the expectation after this change?

LongCatIsLooong committed:
fix segment hit test behavior (#57461)

pedromassango commented:
> Since it is getting more complex, have you thought about write a design doc that summarize the new behavior so that people can quickly understand what is the expectation after this change?

No, but now that you say that, I think it need to be done.

Sorry I never did this before, can you help me on how to do that?

chunhtai commented:
@pedromassango 
You can follow the steps in https://github.com/flutter/flutter/wiki/Chat#design-documents
and then once you are done, you can share it in our #hackers discord channel so people can take a look. Feel free to ping me if you have any question.

lzhuor commented:
Thanks @pedromassango for making this change! This is definitely something many people would love to have.

agordeev committed:
Expect a generic State instead of implementation (#57500)

christopherfujino committed:
[flutter_tools] Add additional bash entrypoint for running dart sdk directly (#57257)

stuartmorgan-g committed:
Remove TRANSFORM from Linux CMake files (#57515)

The Linux CMakeLists.txt are intended to be compatible with 3.10, but
accedintally used a list construct that wasn't added until 3.12. This
adds a custom replacement function.

This makes the build compatible with 3.10 as originally intended.

jonahwilliams committed:
[flutter_tools] chunk the hashing of large files (#57506)

 For files larger than 250 KB (roughly the size of framework.dart), chunk the conversion. This may be important for large assets like images.

jonahwilliams committed:
[flutter_tools] reland: integrate l10n tool into hot reload/restart/build (#57510)

Reland: #56167

jonahwilliams committed:
[flutter_tools] handle missing null check in manifest parser (#57400)

engine-flutter-autoroll committed:
Roll engine 90f45bd3efce..4b1a70e6a256 (4 commits) (#57517)

pedromassango commented:
I've created a Breaking Change proposal file [here](https://docs.google.com/document/d/16HAURcOtYAoHntDOEtdGy9KkKSItulG9Ldyjzdaosqo/edit?usp=sharing).

cg021 committed:
remove pending timers list code out of assert message (#57287)

* move pending timers list code out of assert

* style change

Co-authored-by: Emmanuel Garcia <egarciad@google.com>

* add unit tests

* formatting

* pending timer test

* pending timer assertion test

* add setup and teardown

Co-authored-by: Emmanuel Garcia <egarciad@google.com>

LongCatIsLooong committed:
update appbar documentation (#50673)

jonahwilliams committed:
[flutter_tools] prevent wildcard assets from causing build invalidation issues (#56472)

jonahwilliams committed:
[flutter_tools] remove globals/context for android testing (#57445)

Commit:
Revert "[flutter_tools] remove globals/context for android testing (#57445)" (#57611)

This reverts commit 602d8baf34ec9d875fd6e74e99699bf62885d9f3.

stuartmorgan-g committed:
Remove MaterialControls from examples/flutter_view (#57621)

Co-authored-by: Jenn Magder <magder@google.com>

Piinks committed:
Step 1 of 3: Add opt-in fixing Dialog border radius to match Material Spec (#56084)

justinmc committed:
Fix Web asking for clipboard permissions (#57324)

clocksmith committed:
move paragraph about wide viewports up (#57521)

liyuqian committed:
Reland "Add DevTools memory test (#55486)" (#57340)

darrenaustin committed:
Update the requirements for applying the elevation overlay. (#57526)

Commit:
[flutter_tools] fix incorrect comment on web runner (#57450)

LongCatIsLooong committed:
avoid forceToPoint in hit-testing when possible (#57519)

engine-flutter-autoroll committed:
Roll engine 4b1a70e6a256..2d4e83921d31 (3 commits) (#57529)

jmagman committed:
Always show diagnostics (#57532)

stuartmorgan-g committed:
Re-add line to Linux template CMakeLists.txt (#57538)

Re-adds a line that was accidentally removed in #57515

LongCatIsLooong committed:
Improve CupertinoDatePicker docs (#57534)

jmagman committed:
Add Android private keystore to project gitignore (#57601)

SpajicM commented:
Hi, thanks working on this! I'd like a little clarification - for example, if there's a form, with 5 TextFormField children, does it mean the all 5 fields will "go red" as soon as first one is changed?

Commit:
[flutter_tools] reland: remove globals from android device/testing (#57614)

ashank96 commented:
> Hi, thanks working on this! I'd like a little clarification - for example, if there's a form, with 5 TextFormField children, does it mean the all 5 fields will "go red" as soon as first one is changed?

@SpajicM No, only that formfield goes red whose state will change.

giantsol committed:
Have _warpToCurrentIndex() shortcut logic behave properly (#57574)

* Have _warpToCurrentIndex() shortcut logic behave properly

* Add test

* Apply code review

* Fix typo (fromm -> from)

chunhtai commented:
> Hi, thanks working on this! I'd like a little clarification - for example, if there's a form, with 5 TextFormField children, does it mean the all 5 fields will "go red" as soon as first one is changed?

@pedromassango can you clarify the if this is the case? also it will be good to mention the behavior here in the design document.

pedromassango commented:
> Hi, thanks working on this! I'd like a little clarification - for example, if there's a form, with 5 TextFormField children, does it mean the all 5 fields will "go red" as soon as first one is changed?

Right now it will validate all the form fields. Thinking on this now, maybe I should just validate the one that have its content updated!

pedromassango commented:
> > Hi, thanks working on this! I'd like a little clarification - for example, if there's a form, with 5 TextFormField children, does it mean the all 5 fields will "go red" as soon as first one is changed?
> 
> @pedromassango can you clarify the if this is the case? also it will be good to mention the behavior here in the design document.

Sure

pedromassango committed:
Updated test cases to match new requirements

pedromassango committed:
Make this feature optional for FormField

pedromassango committed:
Added tests for the Form widget

pedromassango committed:
Make this feature opt-in for the Form widget

pedromassango committed:
revert some tests

pedromassango committed:
revert tests related to Dropdown form field

pedromassango committed:
add autovalidateOnUserInteraction param in TextFormField

pedromassango committed:
test that autovalidateOnUserInteraction is passed to super class

pedromassango committed:
Merge branch 'master' of https://github.com/flutter/flutter

pedromassango committed:
Merge branch 'master' into patch-1

pedromassango committed:
Added new API for Form and TextFormField auto-validation

pedromassango committed:
Added autoValidateMode param into DropdownButtonFormField widget for auto-validation

Stonekity committed:
fix memory leak of android view (#50111)

justinmc committed:
Allow updating textAlignVertical (#57033)

perclasson committed:
Add code example for CustomScrollView on how to make it grow in two directions along its scroll axis (#57670)

kenvandine committed:
Desktop default window size (#57274)

engine-flutter-autoroll committed:
Roll Engine from 2d4e83921d31 to 9ce1e5c5c7e7 (27 revisions) (#57629)

dnfield committed:
Functionality to check handlers set on platform channels (#57696)

AyushBherwani1998 committed:
[AppBarTheme] adds centerTitle property (#57736)

jamesderlin committed:
[flutter tools] Improve messages when we fail to connect to the Observatory (#57355)

liyuqian committed:
Use pub inside the Flutter directory (#57704)

pedromassango committed:
Prioritize new parameter for auto validation behaviour

pedromassango committed:
Prevent Form to validate if autoValidateMode was set to onUserInteraction

devoncarew committed:
fix a typo in trace events for the image cache (#57821)

AyushBherwani1998 committed:
[CheckboxListTile] exposes contentPadding property of ListTile. (#57868)

gspencergoog committed:
Update snippets README.md to include more detail (#57152)

Just adding some more details to the snippets README.md file about testing snippets.

pedromassango committed:
Deprecating autovalidate parameter for Form, TextFormField, DropdownTextFormField and FormField widgets.

chunhtai commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -672,9 +672,10 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
     Key key,
     this.onTap,
-    this.value,
+    @required this.value,
```
Is this change required for this pr? If not, let's leave this out because this is probably another breaking change.

chunhtai commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -1466,7 +1467,13 @@ class DropdownButtonFormField<T> extends FormField<T> {
+    @Deprecated(
+        'Use autoValidateMode parameter which provide more specific '
+        'behaviour related to auto validation. '
+        'This feature was deprecated after v1.17.1'
```
the version number doesn't seem accurate. The current tag is 1.19.0

chunhtai replied:
also indent is off

chunhtai commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -1488,7 +1495,9 @@ class DropdownButtonFormField<T> extends FormField<T> {
          validator: validator,
+         // ignore: deprecated_member_use_from_same_package
          autovalidate: autovalidate,
+         autoValidateMode: autoValidateMode,
```
you can do something like
```dart
autoValidateMode: autovalidate ? AutoValidateMode.always : autoValidateMode;
```

chunhtai replied:
so that you don't need to use the deprecated parameter in super class

chunhtai commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -414,6 +414,7 @@ class _DatePickerDialogState extends State<_DatePickerDialog> {
       case DatePickerEntryMode.input:
         picker = Form(
           key: _formKey,
+          // ignore: deprecated_member_use_from_same_package
```
can this use the new api instead?

chunhtai commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -152,6 +152,11 @@ class TextFormField extends FormField<String> {
+    @Deprecated(
+        'Use autoValidateMode parameter which provide more specific '
+            'behaviour related to auto validation. '
+            'This feature was deprecated after v1.17.1'
```
nit:indent and the tag number is inaccurate

pedromassango replied:
Sorry what do you mean by nit indent? this one is what the IDE is suggesting

chunhtai replied:
we do not completely follow the dart style guide, for the deprecation message we always indent by two spaces

pedromassango replied:
Ok

chunhtai commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -183,6 +189,7 @@ class TextFormField extends FormField<String> {
        assert(autocorrect != null),
        assert(enableSuggestions != null),
+       // ignore: deprecated_member_use_from_same_package
        assert(autovalidate != null),
```
also check for autoValidateMode != null

chunhtai commented on packages/flutter/lib/src/material/text_form_field.dart:
```diff
@@ -205,8 +212,10 @@ class TextFormField extends FormField<String> {
+    // ignore: deprecated_member_use_from_same_package
     autovalidate: autovalidate,
     enabled: enabled ?? decoration?.enabled ?? true,
+    autoValidateMode: autoValidateMode,
```
same as above

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -76,10 +76,21 @@ class Form extends StatefulWidget {
+    @Deprecated(
+        'Use autoValidateMode parameter which provide more specific '
+        'behaviour related to auto validation. '
+        'This feature was deprecated after v1.17.1'
```
same as above

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -76,10 +76,21 @@ class Form extends StatefulWidget {
   }) : assert(child != null),
+       autoValidateMode = autoValidateMode != AutoValidateMode.disabled
+           // ignore: deprecated_member_use_from_same_package
+           ? autoValidateMode : autovalidate
```
I will probably simply this so that the autovalidate == true takes the highest priority and use autoValidateMode if autovalidate == false
```dart
autoValidateMode = autovalidate ? AutoValidateMode.always : autoValidateMode;
```

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -125,6 +136,16 @@ class Form extends StatefulWidget {
+  /// Defaults to [AutoValidateMode.disabled] which means no auto-validation
+  /// will occur.
+  ///
+  /// This property has no effect if [autovalidate] is false.
```
This seems incorrect?

pedromassango replied:
Yes. I need to remove this statement:

` /// This property has no effect if [autovalidate] is false.`

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -125,6 +136,16 @@ class Form extends StatefulWidget {
   /// will rebuild.
   final VoidCallback onChanged;
 
+  /// If [AutoValidateMode.onUserInteraction] this form will only auto-validate
```
The first sentence of the doc string should talk about what this property is and for. the detail on how to use this property should come after.

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -283,10 +323,22 @@ class FormField<T> extends StatefulWidget {
+    AutoValidateMode autoValidateMode = AutoValidateMode.disabled,
+  }) :  assert(builder != null),
+        assert(autoValidateMode != null),
+        autoValidateMode = autoValidateMode != AutoValidateMode.disabled
```
same here

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -319,15 +371,30 @@ class FormField<T> extends StatefulWidget {
   /// [FormFieldState.validate] to validate. If part of a [Form] that
   /// auto-validates, this value will be ignored.
+  @Deprecated(
+      'Use autoValidateMode parameter which provide more specific '
```
nit: indent

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -319,15 +371,30 @@ class FormField<T> extends StatefulWidget {
   /// Whether the form is able to receive user input.
   ///
+  // ignore: deprecated_member_use_from_same_package
   /// Defaults to true. If [autovalidate] is true, the field will be validated.
```
This doc should update with the version of autoValidateMode

chunhtai commented on packages/flutter/test/material/dropdown_form_field_test.dart:
```diff
@@ -53,6 +53,7 @@ Widget buildFormFrame({
         child: RepaintBoundary(
           child: DropdownButtonFormField<String>(
             key: buttonKey,
+            // ignore: deprecated_member_use_from_same_package
```
This should update with the new api

chunhtai reviewed (commented):
Thanks for keeping up with this PR! The idea looks solid. I left some comments that are mostly coding style nits.

pedromassango committed:
Improved coding style and updated some properties docs

pedromassango committed:
Removed unnecessary test

pedromassango committed:
Fix indent

pedromassango committed:
fixup! Fix indent

pedromassango committed:
Use new API to prevent analysis error

chunhtai committed:
Improve error message when using popuntil with bad predicate (#57247)

Commit:
Update platform_view.dart (#57516)

Nash0x7E2 committed:
[flutter_tools] Support latest IntelliJ via Jetbrain toolbox (#57963)

jmagman committed:
Remove unused integration test iOS directory (#57624)

jmagman committed:
Allow FLUTTER_APPLICATION_PATH to be null for misconfigured Xcode projects (#57701)

jmagman committed:
Change release archive check to warning (#57688)

jonahwilliams committed:
[flutter_tools] hide all development tools (#57690)

jonahwilliams committed:
[flutter_tools] ensure emulator command does not crash with missing avdmanager (#57703)

jonahwilliams committed:
[flutter_tools] add vm service method to pull SkSL (#57813)

jonahwilliams committed:
[flutter_tools] validate android arch and build number (#57830)

jonahwilliams committed:
[flutter_tools] URI encode dart-define values (#57873)

jonahwilliams committed:
[flutter_tools] throw if asked to build release for x86_64 (#57874)

Commit:
Prevent building non-android plugins in build aar (#58018)

kaparray committed:
Deleted deprecated profile func and profile.dart (#57841)

zanderso committed:
[flutter_tools] Put a heap size limit on the frontend_server (#58039)

justinmc committed:
Consistent American spelling of 'behavior' (#58016)

dnfield committed:
Remove invalid `local` from macos_assemble.sh (#58030)

When you run this, bash complains that `local` is only valid in functions. Gets rid of that extraneous warning.

goderbauer committed:
Link to migration guide tempalte from pull request template (#58027)

engine-flutter-autoroll committed:
Roll Engine from 9ce1e5c5c7e7 to 1a8349888e97 (69 revisions) (#58031)

flar committed:
print checksum differences when detected by --verify-only (#58064)

AyushBherwani1998 committed:
[SwitchListTile] adds controlAffinity property (#58037)

stuartmorgan-g committed:
Fix Linux plugin template build visibility (#58069)

The CMake plugin build wasn't setting visibility to hidden by default,
which meant that plugins exported everything by default. This would make
bad interactions between plugins much more likely; only the intended API
should be exported by the shared library.

Commit:
[flutter_tools] write test to convince self of lack of timing issue (#58011)

In #55864 a race condition was described where a done event is received before we finish connecting. This cannot happen, since async functions begin synchronously and the flag isWaitingForVm is tripped immediately, keeping onDone from exiting.

Commit:
Revert "[flutter_tools] write test to convince self of lack of timing issue (#58011)" (#58123)

This reverts commit f77c669d687c11d6aca81f053bbf028adc59794e.

JoseAlba committed:
Slider value indicator gets disposed if it is activated (#57535)

gspencergoog committed:
Add MediaQueryData.navigationMode and allow controls to be focused when disabled. (#54919)

This adds a new navigationMode to the MediaQueryData that indicates how focusable controls should behave under different navigation modes, currently with two modes: NavigationMode.traditional, and NavigationMode.directional.

It may seem like focusing a disabled control is not desirable, but this is useful for user interfaces that use DPAD navigation because if a control gets disabled, losing focus is disruptive to the user, and it is difficult to control where the focus will end up unless it is done explicitly.

gspencergoog committed:
Update the flutter script's locking mechanism and follow_links (#57590)

Update the flutter and dart scripts' locking mechanism and follow_links function to be more robust and support more platforms.

This adds support for using mkdir as a fallback if the system doesn't have flock instead of using shlock, since shlock doesn't work on shared filesystems.

It also fixes a problem in the follow_links function where it failed when the link resolved to the root directory.

engine-flutter-autoroll committed:
Roll Engine from 1a8349888e97 to 2663be837081 (16 revisions) (#58111)

Commit:
[devicelab] less sensitivity to whitespace in flutter_performance_test testing (#58175)

chunhtai committed:
fix route annoucement for first route and last route (#57339)

engine-flutter-autoroll committed:
Roll Engine from 2663be837081 to 2b616caad05c (2 revisions) (#58136)

onatcipli committed:
fixes isAlwaysShown material scrollbar.dart (#54128)

Commit:
[flutter_tools] forward flutter format to dart format and deprecate (#57829)

Commit:
Update stocks example to use l10n.yaml workflow (#58121)

Commit:
[flutter] allow loading either NOTICES or LICENSE (#58131)

To make #57871 easier to land, add support for loading either LICENSES or NOTICES, preferring the later.

Commit:
[release] remove .dart_tool directory from uploaded zip (#58135)

Fixes #58003

The .dart_tool directory can contain build assets, as well as the package_config.json file - the replacement for .packages. Remove it for the same reason we remove .packages

stuartmorgan-g committed:
Start from a clean slate when bundling Linux build (#58174)

BundleUtilities apparently doesn't do build-system-style timestamp
analysis when deciding what to copy, and instead just doesn't copy
things that are already present. This cleans that bundle directory on
each build, so that it includes the up-to-date library versions.

Since this is just copying from build artifacts, this is very fast; the
build steps themselves are not affected.

Fixes https://github.com/flutter/flutter/issues/58049

stuartmorgan-g committed:
Add release and profile support for Windows (#57749)

Builds and bundles an 'app.so' containing AOT data, and downloads and packages the release/profile artifacts for those builds, rather than always using the debug artifacts.

Fixes https://github.com/flutter/flutter/issues/38477
Fixes https://github.com/flutter/flutter/issues/39664

yjbanov committed:
add a backward variant of BenchCardInfiniteScroll (#58140)

dkwingsmt committed:
Add mouse cursor API to widgets (phase 1) (#57628)

* Adds default cursor and/or mouseCursor property to a number of widgets.
* Adds `MaterialStateMouseCurrsor`.

stuartmorgan-g committed:
Update Windows template version (#58189)

PR #57749 included changes to the way the project interacts with the
build process (e.g., bundling the necessary library), so should have
incremented the template version to trigger the messaging to developers.

engine-flutter-autoroll committed:
Roll Engine from 2b616caad05c to c86dcac156cd (7 revisions) (#58180)

giantsol committed:
Error message when size has not been set in RenderBox's performLayout should be well versed (#58151)

liyuqian committed:
Mark non-flaky test as such (#58112)

Commit:
Revert "[flutter_tools] always initialize the resident runner from dill (#52497)" (#58193)

helin24 committed:
Add function to set structured error early (#58118)

* Add function to set structured error early

* Remove unused imports

* Save test handler in setUp

* Add utils file for shared test service

* Rename structured error functions

* Use setUpAll to save error handler and call initStructuredError from other init

* Rename widget inspector test service

* Remove debugging print statement

* Move error handling setting back to initServiceExtensions

* Rename structured error handler in test

* Namespace environment variable

* Rename test

engine-flutter-autoroll committed:
Roll Engine from c86dcac156cd to c5d012900f7d (2 revisions) (#58214)

engine-flutter-autoroll committed:
Roll Engine from c5d012900f7d to 17737e6fd4ec (6 revisions) (#58221)

pedromassango committed:
Prevent WhitelistingTextInputFormatter to return a empty string if the current value does not satisfy the formatter (#57264)

engine-flutter-autoroll committed:
Roll Engine from 17737e6fd4ec to c19459ab7d3f (2 revisions) (#58230)

dnfield committed:
[e2e] make test bindings friendlier to integration tests (#58210)

Commit:
[flutter_tools] rename output LICENSE file to NOTICES and support loading either (#57871)

Work towards #16723

This is only safe to land after #58131 lands in google3. Only build NOTICES in asset manfiest, and load either LICENSE or NOTICES from pubspec dependencies.

engine-flutter-autoroll committed:
Roll Engine from c19459ab7d3f to 6a470b8bccf0 (3 revisions) (#58255)

gspencergoog committed:
Fix extraneous spaces printed by flutter tool if the lock isn't waited on. (#58215)

This removes the extraneous whitespace printed when the flutter command is invoked, but didn't wait for a lock. It only is a problem when the carriage return isn't honored (as in IDE consoles).

pennzht committed:
Run `flutter update-packages --force-upgrade`. (#58104)

* Run `flutter update-packages --force-upgrade`.
* Update test to 1.14.5.

MarcusTomlinson committed:
[flutter_tools] Support profile and release builds on Linux (#57135)

liyuqian committed:
Add clipBehavior to widgets with clipRect (#55977)

* Add clipBehavior to RenderFlex

* Add clipBehavior to FittedBox

* Add clipBehavior to Flex and FittedBox

* Add clipBehavior to UnconstrainedBox

* Add clipBehavior to Stack and Wrap

* Add clipBehavior to TextEditable

* Add clipBehavior to ListWheelScrollView

* Add clipBehavior to SingleChildScrollView

* Add clipBehavior to RenderViewportBase's widgets

Those widgets are NestedScrollView and ShrinkWrappingViewport.

* Fix tests

* Remove enum Overflow and fix typo

* Remove clipToSize

* Analyze fix

* Remove Mixin and other small fixes

* Fix tests and respect Stack's default clipBehavior

* Add Overflow back to make it non-breaking

* Restore clipBehavior to make it non-breaking

* Small fixes

* Fix rebase

Commit:
Revert "Revert "[flutter_tools] always initialize the resident runner from di…" (#58208)

justinmc committed:
Helpful assertion for isAlwaysShown error (#58258)

jmagman committed:
Detect USB/network interface from iOS devices (#58257)

gspencergoog committed:
Remove callback asserts on FocusableActionDetector (#58272)

This makes the callback arguments to FocusableActionDetector optional, if you (for instance) only want to define shortcuts and actions and a focus node for something.

christopherfujino committed:
[flutter_tools] remove execute permissions on shared.sh, and move off path (#58267)

engine-flutter-autoroll committed:
Roll Engine from 6a470b8bccf0 to 4d78121a11db (3 revisions) (#58276)

engine-flutter-autoroll committed:
685699b35 Roll Fuchsia Linux SDK from wmNGH... to in0Aq... (flutter/engine#18671) (#58282)

Hixie committed:
Dedupe (and update) the --track-widget-creation documentation (#58206)

engine-flutter-autoroll committed:
Roll Engine from 685699b351a9 to 3dfbe722b965 (2 revisions) (#58293)

flar committed:
Enabling the ImageFiltered(ImageFilter.matrix) page of macrobenchmark (#58277)

engine-flutter-autoroll committed:
Roll Engine from 3dfbe722b965 to 923687f0e7ff (2 revisions) (#58340)

mehmetf committed:
Revert "Add clipBehavior to widgets with clipRect (#55977)" (#58344)

This reverts commit cd593dae1961ca2bc5c045a310f16e6cdffc8294.

pedromassango committed:
Revert: deprecate autovalidate param

pedromassango committed:
fixup! Revert: deprecate autovalidate param

chunhtai commented on packages/flutter/lib/src/material/dropdown.dart:
```diff
@@ -1488,7 +1489,8 @@ class DropdownButtonFormField<T> extends FormField<T> {
          initialValue: value,
          validator: validator,
-         autovalidate: autovalidate,
+         // ignore: deprecated_member_use_from_same_package
```
lets remove this for now

chunhtai replied:
here and other places

pedromassango replied:
You mean remove the autovalidate parameter? But what if someone used the FormField Widget directly?

chunhtai replied:
I mean the `// ignore: deprecated_member_use_from_same_package` and reintroduce it in the deprecation PR later

pedromassango replied:
Hahah. Right, I misunderstood you

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -125,6 +129,19 @@ class Form extends StatefulWidget {
   /// will rebuild.
   final VoidCallback onChanged;
 
+  /// Used to enable form fields auto validation and update their error text.
```
This sentence is a bit inaccurate, because it can disable auto validate as well

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -125,6 +129,19 @@ class Form extends StatefulWidget {
 
+  /// Used to enable form fields auto validation and update their error text.
+  ///
+  /// If [AutoValidateMode.onUserInteraction] this form will only auto-validate
```
This should also mention the disable

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -137,13 +154,17 @@ class Form extends StatefulWidget {
     if (widget.onChanged != null)
       widget.onChanged();
+    for (final FormFieldState<dynamic> field in _fields) {
+      _hasInteractedByUser = field._hasInteractedByUser == true;
```
This seems wrong, only the last one will affect the result. Also you should be able to exit early if one of them is true, you can do _fields.any(..)

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -323,11 +361,25 @@ class FormField<T> extends StatefulWidget {
 
+  /// Used to enable auto validation and update their error text.
+  ///
+  /// If [AutoValidateMode.onUserInteraction] this form will only auto-validate
```
This form field?

stuartmorgan-g committed:
Fix non-local-engine Linux release builds (#58372)

kittykat committed:
Fix typo in error message in flutter doctor (#58421)

gspencergoog committed:
Revert flutter command to shlock when flock isn't available (#58281)

This reverts the flutter command to use shlock when flock isn't available. It seems that the mkdir method isn't as reliable as we want. I think that this is because the trap isn't always be executed, which is why I think that shlock uses PIDs to help it be more reliable. Unfortunately, that means that we're back to not working over network shares (which is where things were before I moved to the mkdir method, so not really a regression). I did leave in the mkdir method for platforms that have neither flock nor shlock (which should be very few and far between, but still), so at least we'll do some locking there now.

LongCatIsLooong committed:
EditableText.bringIntoView calls showOnScreen (#58346)

jmagman committed:
Change iOS device discovery from polling to long-running observation (#58137)

Commit:
[flutter_driver] make timeline request in chunks (#58430)

Work-around large timeline data killing devicelab machines by requesting data in 1 second intervals and combining at the end. Non-breaking change to the driver API.

engine-flutter-autoroll committed:
Roll Engine from 923687f0e7ff to d501c49cce4c (5 revisions) (#58436)

Commit:
[flutter_tools] cleanup to devfs Operations (#58332)

removes globals from devFS operations and cleans up testing to be context free. Removes unused and unnecessarily temp directory in testing.

engine-flutter-autoroll committed:
c5234bce6 Roll Skia from 8c5786622b4e to 502eab4fc913 (2 revisions) (flutter/engine#18707) (#58443)

Commit:
[flutter_tools] do not include material icon incorrectly (#58335)

If a dependency specified uses-material-design: true and the main pubspec specifies uses-material-design: false, then the MaterialIcons font would be included in the font manifest, but not in the AssetManifest or final bundle. Remove it from the FontManifest if this occurs

Commit:
[flutter_tools] only copy cached dill after startup (#58188)

Remove copying from the shutdown stage since that seems risky. If the tool copies the first compilation there will still be a decent dill for initialization.

engine-flutter-autoroll committed:
Roll Engine from c5234bce6e9f to e39301f23f32 (2 revisions) (#58449)

Commit:
Revert "[flutter_tools] only copy cached dill after startup (#58188)" (#58454)

This reverts commit 39d1e4b7e9222394d6013aa21731968b36b30cc7.

pedromassango committed:
Fix possible issue with user interaction verification code

pedromassango committed:
Update comment docs and use the new API for test cases

jmagman committed:
Remove outdated disable_input_output_paths from example project Podfiles (#58444)

engine-flutter-autoroll committed:
Roll Engine from e39301f23f32 to 5a085ac3c6a2 (6 revisions) (#58460)

VisheshAtWork committed:
Added Documentation for AnimationController.repeat() (#58098)

chunhtai commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +584,215 @@ void main() {
+    expect(formFieldState.hasError, isTrue);
+  });
+
+  testWidgets('Form auto-validate form fields only after one of them changes if autovalidateMode is onUserInteraction', (WidgetTester tester) async {
```
In this test can you have more than one FormField and verify it starts to validate all FormFields after one of them has changed?

chunhtai commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +584,215 @@ void main() {
+    await tester.pump();
+    expect(find.text(errorText('bar')), findsOneWidget);
+
+    // resetting the form state should remove the error text
```
Same here

chunhtai commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +584,215 @@ void main() {
+      );
+    }
+
+    // The issue only happens on the second build so we need to rebuild the three
```
```suggestion
    // The issue only happens on the second build so we need to rebuild the three.
```

pedromassango committed:
Update packages/flutter/lib/src/widgets/form.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

jmagman committed:
Revert "Remove outdated disable_input_output_paths from example project Podfiles (#58444)" (#58504)

This reverts commit 8adaea90833418c8eac8f3ff103ab9b38006b187.

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: chunhtai <47866232+chunhtai@users.noreply.github.com>

dnfield committed:
Do not assume imageCache is created when handleMemoryPressure is called (#58503)

* Do not assume imageCache is created when handleMemoryPressure is called

* comment for test

Commit:
[devicelab] mark ios transition_perf as non-flaky (#58499)

jmagman committed:
Rename integration test ios_app_with_watch_companion -> ios_app_with_extensions (#58458)

dkwingsmt committed:
InkWell uses MaterialStateMouseCursor and defaults to clickable (#58448)

* Adds support for material state mouse cursor to InkWell and InkResponse
* Set their default cursors to be MaterialStateMouseCursor.clickable.

pedromassango commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -97,6 +97,7 @@ void main() {
               child: Material(
                 child: Form(
                   key: formKey,
+                  // ignore: deprecated_member_use_from_same_package
```
Yes

pedromassango commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +584,215 @@ void main() {
+
+    await tester.pumpWidget(builder());
+
+    // no error text is visible yet
```
I need to get used to this :)

dkwingsmt committed:
Add doc and test for Container's hitTest behavior (#57522)

engine-flutter-autoroll committed:
Roll Engine from 5a085ac3c6a2 to 2a024ead01e7 (5 revisions) (#58509)

pedromassango committed:
Updated test case to make sure Form widget validates all form fields when one changes

Hixie committed:
Various bits of trivial cleanup (#58203)

pedromassango committed:
Uses the new API and fixed typo

pedromassango committed:
Merge branch 'master' of https://github.com/flutter/flutter into patch-1

chunhtai commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +578,232 @@ void main() {
+      );
+    }
+
+    // The issue only happens on the second build so we need to rebuild the three.
```
This sentence is a bit confusing to people who do not know the context, maybe `Makes sure it won't autovalidate after rebuilds if there is no user interaction`

pedromassango replied:
Here I'm just trying to explain why I called pumpWidget twice

chunhtai replied:
I understand, other people might not know what is the issue you are referring here. Can you add a sentence or two to explain it?

pedromassango replied:
Right

pedromassango replied:
> This sentence is a bit confusing to people who do not know the context, maybe `Makes sure it won't autovalidate after rebuilds if there is no user interaction`

I will replace the test desciption with your sentence since its more clear for the users

chunhtai reviewed (approved):
LGTM with a minor nit

pedromassango committed:
Improve test comments

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,10 +479,34 @@ class FormFieldState<T> extends State<FormField<T>> {
+
+/// Used to configure the auto validation of [TextFormField] and [Fom] widgets.
+enum AutoValidateMode {
+  /// no auto validation will occur.
```
capital letter

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,10 +479,34 @@ class FormFieldState<T> extends State<FormField<T>> {
+  /// no auto validation will occur.
+  disabled,
+
+  /// used to auto-validate [Form] and [FormField] even without user interaction.
```
Same here

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -430,10 +479,34 @@ class FormFieldState<T> extends State<FormField<T>> {
+  /// used to auto-validate [Form] and [FormField] even without user interaction.
+  always,
+
+  /// used to auto-validate [Form] and [FormField] only after each user
```
and here

chunhtai reviewed (commented):
I missed this one in previous review. This should be the last nit I have

justinmc reviewed (commented):
Just adding some improvements to docs and test descriptions here.

pedromassango committed:
Update packages/flutter/lib/src/widgets/form.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/lib/src/widgets/form.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/lib/src/widgets/form.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +578,233 @@ void main() {
+    }
+
+    // The issue only happens on the second build so we
+    // need to rebuild the three twice.
```
Yes

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango committed:
Update packages/flutter/test/widgets/form_test.dart

Co-authored-by: Justin McCandless <justinjmccandless@gmail.com>

pedromassango commented on packages/flutter/test/widgets/form_test.dart:
```diff
@@ -578,4 +578,233 @@ void main() {
+    expect(find.text(errorText('')), findsOneWidget);
+  });
+
+  testWidgets('autovalidate parameter is still used if true', (WidgetTester tester) async {
```
Actualy no. This test assert that the old parameter is still used. I don't know if this still a valid test

pedromassango replied:
I'm thinking to update it to use the old parameter and remove this test case once the old parameter is removed.

pedromassango replied:
Or I can just delete this test case. 

Cc: @chunhtai

pedromassango committed:
Update test case to use the old parameter

chunhtai commented:
@pedromassango the CLA seems unhappy, can you sign it agian?

pedromassango commented:
> @pedromassango the CLA seems unhappy, can you sign it again?

Working on this. I signed one of the PR with another email account. Now my boss need to recover it so I can sign the CLA.

pedromassango commented:
@googlebot  I fixed it

pedromassango commented:
@googlebot I fixed it.

pedromassango commented:
@googlebot I fixed it.

pedromassango commented:
@googlebot I fixed it.

pedromassango commented:
Due to CLA problems I've moved this work at https://github.com/flutter/flutter/pull/59766.
